src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/activities/ActivitiesService.kt

```diff
@@ -76,6 +77,43 @@ class ActivitiesService(
     }
   }
 
+  fun createAllocation(allocationEvent: AllocationDomainEvent) {
+    activitiesApiService.getActivitySchedule(allocationEvent.scheduleId).let { schedule ->
+      activitiesApiService.getAllocation(allocationEvent.allocationId).let { allocation ->
+        mappingService.getMappingGivenActivityScheduleId(allocationEvent.scheduleId).let { mapping ->
+
+          val telemetryMap = mutableMapOf(
+            "activityScheduleId" to schedule.id.toString(),
+            "allocationId" to allocation.id.toString(),
+            "offenderNo" to allocation.prisonerNumber,
+            "bookingId" to allocation.bookingId.toString(),
+            "prisonId" to schedule.activity.prisonCode,
+          )
+
+          val nomisResponse = try {
+            nomisApiService.createAllocation(
+              mapping.nomisCourseActivityId,
+              CreateOffenderProgramProfileRequest(
+                bookingId = allocation.bookingId!!,
+                startDate = allocation.startDate,
+                endDate = allocation.endDate
+              )
+            )
+          } catch (e: Exception) {
+            telemetryClient.trackEvent("activity-allocation-create-failed", telemetryMap)
+            log.error("createAllocation() Unexpected exception", e)
+            throw e
+          }
+
+          val mapWithNomisId = telemetryMap
+            .plus(Pair("offenderProgramReferenceId", nomisResponse.offenderProgramReferenceId.toString()))
+
+          telemetryClient.trackEvent("activity-allocation-created-event", mapWithNomisId)
+        }
+      }
+    }
+  }
```

Comment:
not sure about whether an allocation <-> offender program profile mapping is needed. I didnt think so, but possibly a prisoner could be allocated the same course twice ?

Reply:
Don't we need it to deallocate a prisoner from a course?

Reply:
OFFENDER_PROGRAM_PROFILES has a unique index on :
  OFFENDER_BOOK_ID, 
  CRS_ACTY_ID, 
  DECODE("OFFENDER_PROGRAM_STATUS",'ALLOC','ALLOC',TO_CHAR("OFF_PRGREF_ID"))
so the OPP row can always be found from activityschedule -> courseActivity and booking id provided the allocation is active (ALLOC).
I dont think we need mapping the other way for 1 way sync